Source/Proxy/IPublicCallContext.cs

```diff
@@ -0,0 +1,21 @@
+using System.Reflection;
+
+namespace Moq.Proxy
+{
+	/// <summary>
+	/// Class that represents the method call currently being executed. 
+	/// Allows for custom interception for open generic methods
+	/// </summary>
+	public interface IPublicCallContext
+	{
+		/// <summary>
+		/// Arguments passed into the method
+		/// </summary>
+		object[] Arguments { get; }
+
+		/// <summary>
+		/// Closed generic method currently being executed.
+		/// </summary>
+		MethodInfo Method { get; }
+	}
+}
```

Comment:
* I understand now why you chose to make part of the original interface public. (Even though it won't allow you to reliably discover the generic parameter types: You cannot call `.GetType()` on an argument that is `null`!) So far so good. What I am not entirely sure about is how your change potentially affects the preexisting "separation of concerns" between `.Callback()` and `.Returns()`. Your new `.Returns()` overload can be a little bit of both. This makes me wonder whether the new `.Returns()` is more similar to, and therefore should be, a `.Callback()` instead. I really don't mind the new syntax at all, I just can't foresee what consequences this might have further down the road.

* Splitting the interface to keep internals hidden is fine with me. I also think that you've made a good choice for what should go into the public interface. As long as noone demands `InvokeBase` or `SetArgumentValue`, let's keep them internal. They can always be added later, but once they're part of the public API, it'll be be more difficult to take them away again.

* Yes, please move the public interface to the `Moq` namespace. `Moq.Proxy` appears to bundle a few internal things surrounding DynamicProxy, and the end user probably doesn't care about that.

* Regarding the interfaces' names, to me it seems somewhat counterintuitive to let the internal `ICallContext` inherit from the public `IPublicCallContext`. The base interface's name would appear to imply that `ICallContext` has to be public, too. I personally find it more logical to have a `public interface ICallContext` (equivalent to your `IPublicCallContext`) and to have an internal subtype, `internal interface IExtendedCallContext` or something (which is today's `ICallContext`, but hopefully you can find a better new name for it ;-).